File: PROGRAMACION-2/repaso-prueba/banco.cpp
//banco abrir y cerrar cuentas, depositar y sacar plata

#include <iostream>
#include <vector>

class Persona
{
    public:
    //atributo
    std::string nombre;

    //constructor 
    Persona (std::string nombre) : nombre(nombre) {}
};

class Banco
{
    private:
    std::vector <int> saldos;
    std::vector <std::string> clientes;

    public:

    void crear_cuenta(std::string nombre, int saldo_inicial=0)
    {
        if (saldo_inicial == 0)
        {
            saldos.push_back(saldo_inicial);
            clientes.push_back(nombre);
        }

        else
        {
            saldos.push_back(saldo_inicial);
            clientes.push_back(nombre);
        }
    }

    void eliminar_cuenta(std::string nombre)
    {
        size_t tam = size(clientes);

        for (int i=0; i<tam; i++)
        {
            if (clientes.at(i) == nombre)
            {
                clientes.at(i) = "vacio";
                saldos.at(i) = 0;
            }
        }
    }

    void agregar_saldo(int id, int saldo_agregar)
    {
        saldos.at(id) += saldo_agregar;
        std::cout << clientes.at(id) << " tu nuevo saldo es de: " << saldos.at(id) << std::endl;
    }

    void sacar_saldo(int id, int saldo_sacar)
    {
        if (saldos.at(id) <= 0)
        {
            std::cout << clientes.at(id) << " accion invalida tu saldo es de: " << 0 << std::endl;
        }

        else if (saldos.at(id) > 0 && saldos.at(id) - saldo_sacar <= 0)
        {
            std::cout << clientes.at(id) << " lo sentimos el monto a sacar " << saldo_sacar << " es demasiado grande" << std::endl;
            std::cout << clientes.at(id) << " actualmente tienes un saldo de: " << saldos.at(id) << std::endl;
        }

        else if (saldos.at(id) > 0 && saldos.at(id) - saldo_sacar >= 0)
        {
            saldos.at(id) -= saldo_sacar;

            std::cout << clientes.at(id) << " te nuevo saldo es de: " << saldos.at(id) << std::endl;
        }
    }

    void mostrar_datos_cliente()
    {
        size_t tam = size(clientes);

        for (int i=0; i<tam; i++)
        {
            if (clientes.at(i) == "vacio")
            {
                std::cout <<"id: " << i << " el cliente elimino su cuenta" << std::endl;
            }

            else
            {
            std::cout <<"id: " << i << " cliente: " << clientes.at(i);
            std::cout <<" saldo: " << saldos.at(i) << std::endl;   
            }
        }
    }
};

int main()
{
    Persona p1("oscar");
    Persona p2("felipe");
    Persona p3("fernando");

    Banco chile;

    chile.crear_cuenta(p1.nombre, 100);
    chile.crear_cuenta(p2.nombre, 200);
    chile.crear_cuenta(p3.nombre);
    chile.mostrar_datos_cliente();
    std::cout << std::endl;

    chile.eliminar_cuenta("oscar");
    chile.mostrar_datos_cliente();
    std::cout << std::endl;

    chile.agregar_saldo(1, 100);
    std::cout << std::endl;

    chile.sacar_saldo(1, 250);
    chile.sacar_saldo(2, 500);


    return 0;
}

File: PROGRAMACION-2/repaso-prueba/banco.py
class Persona:
    def __init__(self, nombre):
        self.nombre = nombre

class Banco:
    def __init__(self):
        #almacena un id y el nombre de la persona/cliente
        self.clientes_activos = []
        #almacena el nombre y el saldo del cliente
        self.cuentas_activas = []

    #creacion de las cuentas
    def crear_cuenta(self,cliente, saldo=0):
        if saldo == 0:
            self.cuentas_activas.append([cliente.nombre, saldo])
            self.clientes_activos.append(cliente)
            print(f"{cliente.nombre} has creado tu cuenta con exito y tienes un saldo inicial de 0")

        else:
            self.cuentas_activas.append([cliente.nombre, saldo])
            self.clientes_activos.append(cliente)
            print(f"{cliente.nombre} has creado tu cuenta con exito y tienes un saldo inicial de {saldo}")

    #eliminacion de las cuentas
    def eliminar_cuenta(self, cliente):
        for i in range(0, len(self.cuentas_activas)):
            if cliente in self.clientes_activos and cliente.nombre == self.cuentas_activas[i][0]:
                self.cuentas_activas.remove(self.cuentas_activas[i])
                self.clientes_activos.remove(cliente)
                print(f"has eliminado del cliente {cliente.nombre}")

    #agrega saldo a la cuenta del banco
    def agragar_saldo(self):
        pass

    #sacar saldo a la cuenta del banco
    def sacar_saldo(self):
        pass

    #muestra las cuentas que estan en "self.cuentas_activas"
    def mostrar_cuentas(self):
        print(self.cuentas_activas)

    #muestra los clientes que estan en "self.clientes_activos"
    def mostrar_clientes(self):
        indice = 0
        for cliente in self.clientes_activos:
            indice +=1
            print(f"cliente {indice}: {cliente.nombre}")

p1 = Persona("oscar")
p2 = Persona("felipe")

main = Banco()

print("cuentas iniciales")
main.mostrar_cuentas()
input("")

print("creacion de las cuentas")
main.crear_cuenta(p1)
main.crear_cuenta(p2, 500)
input("")

print("muestra las cuentas activas")
main.mostrar_cuentas()
input("")

print("muestra los clientes activos")
main.mostrar_clientes()
input("")

print("eleminacion de las cuentas")
main.eliminar_cuenta(p1)
main.mostrar_cuentas()
input("")

print("eleminacion de las cuentas")
main.eliminar_cuenta(p2)
main.mostrar_cuentas()